keps/sig-scheduling/2044-topology-awareness-in-kube-scheduler/README.md

```diff
@@ -0,0 +1,416 @@
+nodes in the cluster needs to be exposed to the scheduler so that it can use it to make a more
+informed scheduling decision. This KEP describes how it would be implemented.
+
+## Goals
+
+- Make scheduling process more precise when we have NUMA topology on the
+worker node.
+- Enhance the node object to capture topology information which can be referred to 
+by the scheduler.
+
+## Non-Goals
+
+- Change the PodSpec to allow requesting a specific node topology manager policy
+- This Proposal requires exposing NUMA topology information. This KEP doesn't
+describe how to expose all necessary information it just declare what kind of
+information is necessary.
+- Changes to the TopologyManager and its policies.
+
+# Proposal
+
+Kube-scheduler plugin will be moved from kuberntes-sigs/scheduler-plugin (or out-of-tree)
+plugin into the main tree as a built-in plugin. This plugin implements a simplified version of Topology Manager and hence is different from original topology manager algorithm. Plugin would
+be disabled by default and when enabled would check for the ability to run pod only in case of single-numa-node policy on the node, since it is the most strict policy, it implies that the launch on the node with other existing policies will be successful if the condition for single-numa-node policy passed for the worker node.
+
+To work, this plugin requires topology information of the available resource on the worker nodes.
+
+Kubelet will be responsible for collecting all necessary resource information of the pods,
+based on allocatable resources on the node and allocated resources to pods. The NUMA nodes
+would be represented as Zones in Kubelet and the NodeResourceTopology would capture the 
+resource information at a zone level granularity.
+
+Once the information is captured in the NodeResourceTopology API, the scheduler can refer to
+it like it refers to Node Capacity and Allocatable while making a Topology-aware Scheduling decision.
+
+
+## Changes to the API
+
+Code responsible for working with NodeResourceTopology API will be placed in the stagingit g directory
+at path staging/src/k8s.io/api/node/v1/types.go.
+
+```go
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// NodeResourceTopology is a specification for a hardware resources
+type NodeResourceTopology struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	TopologyPolicy []string `json:"topologyPolicies"`
+	Zones          ZoneMap  `json:"zones"`
+}
+
+// Zone is the spec for a NodeResourceTopology resource
+type Zone struct {
+    Name       string           `json:"name"`
+    Type       string           `json:"type"`
+    Parent     string           `json:"parent,omitempty"`
+    Costs      CostList         `json:"costs,omitempty"`
+    Attributes AttributeList    `json:"attributes,omitempty"`
+    Resources  ResourceInfoList `json:"resources,omitempty"`
+}
```

Comment:
Adding documentation comments for each of the attributes would be extremely helpful as well. 

Adding doc comments to ResourceInfo, CostInfo, and AttributeInfo should be added also since they would be public structs.

Reply:
I have provided how these fields are populated here: https://github.com/kubernetes/enhancements/pull/2787/files#r666352478. Maybe this should be made clearer.

Reply:
better also provide the comments

Reply:
+1 for comments
Just by reading the examples, I still don't understand what is costs.